<img src='https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQ-VfNtOyJbsaxu43Kztf_cv1mgBG6ZIQZEVw&usqp=CAU'>

# Procesamiento de Lenguaje Natural
## Taller #5: TF-IDF

In [2]:
import re
import numpy as np
import pandas as pd

from nltk.corpus import stopwords
stopwords_sp = stopwords.words('spanish')

from nltk.stem.snowball import SnowballStemmer
spanishStemmer=SnowballStemmer("spanish")

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_distances

# Punto 1: Pre-Procesamiento

- `[9 pts]` Leer el archivo `bob_esponja.csv` usando `pandas`
- `[9 pts]` Crear una nueva columna con el texto en minúscula, sin caracteres especiales ni números, sin palabras vacias y hacer stemming de las palabras

In [131]:
BOB = pd.read_csv("C:/Users/Aleja/Documents/Universidad/NLP/Taller 5/bob_esponja.csv", sep=',')

In [117]:
BOB.head(10)

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  
0  Bob Esponja es una esponja de mar con forma re...  
1  Es una estrella de mar de color rosa, el mejor...  
2  Calamardo Tentáculos es un personaje principal...  
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...  
4  Don Eugenio H. Cangrejo (normalmente llamado S...  
5  Sheldon J. Plankton, o simplemente Plankton, e...  
6  Karen Plankton es uno de los dos principales a...  
7  Perla "Perlita" Cangrejo es un personaje princ...  
8  La Señora Puff es un personaje principal de Bo...  
9  Gary el Caracol es un personaje principal de B...

In [132]:
def pre_procesado(texto):
    texto = texto.lower()
    texto = re.sub(r"[\W\d_]+", " ", texto)
    texto = [palabra for palabra in texto.split() if palabra not in stopwords_sp]
    texto = [spanishStemmer.stem(palabra) for palabra in texto]
    texto = " ".join(texto)
    return texto


BOB['pre-procesado'] = BOB['Descripción'].apply(lambda texto: pre_procesado(texto))

BOB


Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  \
0  Bob Esponja es una esponja de mar con forma re...   
1  Es una estrella de mar de color rosa, el mejor...   
2  Calamardo Tentáculos es un personaje principal...   
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...   
4  Don Eugenio H. Cangrejo (normalmente llamado S...   
5  Sheldon J. Plankton, o simplemente Plankton, e...   
6  Karen Plankton es uno de los dos principales a...   
7  Perla "Perlita" Cangrejo es un personaje princ...   
8  La Señora Puff es un personaje principal de Bo...   
9  Gary el Caracol es un personaje principal de B...   

                                       pre-procesado  
0  bob esponj esponj mar form rectangul color ama...  
1  estrell mar color ros mejor amig bob esponj ju...  
2  calamard tentacul personaj principal bob espon...  
3  arenit mejill sandy cheeks ingles personaj pri...  
4  don eugeni h cangrej normal llam señor cangrej...  
5  sheldon j plankton simplement plankton dos ant...  
6  kar plankton dos principal antagon bob esponj ...  
7  perl perlit cangrej personaj principal bob esp...  
8  señor puff personaj principal bob esponj maest...  
9  gary caracol personaj principal bob esponj que...

# Punto 2: TF-IDF

- `[16 pts]` Crear la matriz TF-IDF

In [133]:
BOB.set_index('Personaje', inplace= True) 
BOB.index.name = None
BOB


Descripción  \
Bob Esponja Pantalones Cuadrados  Bob Esponja es una esponja de mar con forma re...   
Patricio Estrella                 Es una estrella de mar de color rosa, el mejor...   
Calamardo Tentáculos              Calamardo Tentáculos es un personaje principal...   
Arenita Mejillas                  Arenita Mejillas (Sandy Cheeks en inglés) en u...   
Eugenio H. Cangrejo               Don Eugenio H. Cangrejo (normalmente llamado S...   
Sheldon J. Plankton               Sheldon J. Plankton, o simplemente Plankton, e...   
Karen Plankton                    Karen Plankton es uno de los dos principales a...   
Perlita Cangrejo                  Perla "Perlita" Cangrejo es un personaje princ...   
Sra. Puff                         La Señora Puff es un personaje principal de Bo...   
Gary el Caracol                   Gary el Caracol es un personaje principal de B...   

                                                                      pre-procesado  
Bob Esponja Pantalones Cuadrados  bob esponj esponj mar form rectangul color ama...  
Patricio Estrella                 estrell mar color ros mejor amig bob esponj ju...  
Calamardo Tentáculos              calamard tentacul personaj principal bob espon...  
Arenita Mejillas                  arenit mejill sandy cheeks ingles personaj pri...  
Eugenio H. Cangrejo               don eugeni h cangrej normal llam señor cangrej...  
Sheldon J. Plankton               sheldon j plankton simplement plankton dos ant...  
Karen Plankton                    kar plankton dos principal antagon bob esponj ...  
Perlita Cangrejo                  perl perlit cangrej personaj principal bob esp...  
Sra. Puff                         señor puff personaj principal bob esponj maest...  
Gary el Caracol                   gary caracol personaj principal bob esponj que...

In [138]:
tfidf_vect = TfidfVectorizer()
tfidf = tfidf_vect.fit_transform(BOB['pre-procesado'])

tfidf_matrix = pd.DataFrame(tfidf.toarray(), columns=tfidf_vect.get_feature_names())
tfidf_matrix.index = BOB.index

tfidf_matrix=tfidf_matrix.T.round(3)
tfidf_matrix

Bob Esponja Pantalones Cuadrados  Patricio Estrella  \
abaj                                 0.057               0.00   
aborrec                              0.000               0.00   
acab                                 0.000               0.00   
accident                             0.049               0.00   
acept                                0.000               0.00   
...                                    ...                ...   
viv                                  0.000               0.05   
vol                                  0.057               0.00   
voz                                  0.195               0.00   
vuelv                                0.172               0.00   
zapat                                0.043               0.00   

          Calamardo Tentáculos  Arenita Mejillas  Eugenio H. Cangrejo  \
abaj                     0.000             0.000                0.000   
aborrec                  0.091             0.000                0.000   
acab                     0.091             0.000                0.000   
accident                 0.000             0.000                0.045   
acept                    0.000             0.000                0.000   
...                        ...               ...                  ...   
viv                      0.060             0.108                0.070   
vol                      0.000             0.000                0.000   
voz                      0.000             0.139                0.000   
vuelv                    0.000             0.000                0.000   
zapat                    0.000             0.000                0.000   

          Sheldon J. Plankton  Karen Plankton  Perlita Cangrejo  Sra. Puff  \
abaj                      0.0           0.000               0.0      0.000   
aborrec                   0.0           0.000               0.0      0.000   
acab                      0.0           0.000               0.0      0.000   
accident                  0.0           0.000               0.0      0.000   
acept                     0.0           0.064               0.0      0.000   
...                       ...             ...               ...        ...   
viv                       0.0           0.000               0.0      0.000   
vol                       0.0           0.000               0.0      0.000   
voz                       0.0           0.000               0.0      0.000   
vuelv                     0.0           0.000               0.0      0.000   
zapat                     0.0           0.000               0.0      0.093   

          Gary el Caracol  
abaj                0.000  
aborrec             0.000  
acab                0.000  
accident            0.000  
acept               0.000  
...                   ...  
viv                 0.000  
vol                 0.000  
voz                 0.000  
vuelv               0.000  
zapat               0.068  

[709 rows x 10 columns]

# Punto 3: Distancia del coseno
- `[12 pts]` Calcular la distancia del coseno entre cada una de los personajes
- `[2 pts]` ¿Cuáles son los personajes más parecidos?
- `[2 pts]` ¿Cuáles son los personajes más diferentes?


In [140]:
dist_cos = cosine_distances(tfidf_matrix.T.values)
dist_cos = pd.DataFrame(dist_cos, columns = tfidf_matrix.columns, index = tfidf_matrix.columns)
dist_cos

Bob Esponja Pantalones Cuadrados  \
Bob Esponja Pantalones Cuadrados                          0.000000   
Patricio Estrella                                         0.826052   
Calamardo Tentáculos                                      0.918644   
Arenita Mejillas                                          0.838317   
Eugenio H. Cangrejo                                       0.868204   
Sheldon J. Plankton                                       0.872076   
Karen Plankton                                            0.938609   
Perlita Cangrejo                                          0.885810   
Sra. Puff                                                 0.858077   
Gary el Caracol                                           0.924158   

                                  Patricio Estrella  Calamardo Tentáculos  \
Bob Esponja Pantalones Cuadrados           0.826052              0.918644   
Patricio Estrella                          0.000000              0.790530   
Calamardo Tentáculos                       0.790530              0.000000   
Arenita Mejillas                           0.841122              0.919857   
Eugenio H. Cangrejo                        0.831634              0.831344   
Sheldon J. Plankton                        0.966449              0.917606   
Karen Plankton                             0.930045              0.949913   
Perlita Cangrejo                           0.891502              0.967265   
Sra. Puff                                  0.952583              0.968111   
Gary el Caracol                            0.894790              0.851644   

                                  Arenita Mejillas  Eugenio H. Cangrejo  \
Bob Esponja Pantalones Cuadrados          0.838317             0.868204   
Patricio Estrella                         0.841122             0.831634   
Calamardo Tentáculos                      0.919857             0.831344   
Arenita Mejillas                          0.000000             0.908202   
Eugenio H. Cangrejo                       0.908202             0.000000   
Sheldon J. Plankton                       0.967774             0.687380   
Karen Plankton                            0.948323             0.891897   
Perlita Cangrejo                          0.962015             0.796723   
Sra. Puff                                 0.972120             0.896080   
Gary el Caracol                           0.900641             0.920664   

                                  Sheldon J. Plankton  Karen Plankton  \
Bob Esponja Pantalones Cuadrados             0.872076        0.938609   
Patricio Estrella                            0.966449        0.930045   
Calamardo Tentáculos                         0.917606        0.949913   
Arenita Mejillas                             0.967774        0.948323   
Eugenio H. Cangrejo                          0.687380        0.891897   
Sheldon J. Plankton                          0.000000        0.794981   
Karen Plankton                               0.794981        0.000000   
Perlita Cangrejo                             0.922577        0.966088   
Sra. Puff                                    0.986733        0.984919   
Gary el Caracol                              0.969239        0.965252   

                                  Perlita Cangrejo  Sra. Puff  Gary el Caracol  
Bob Esponja Pantalones Cuadrados          0.885810   0.858077         0.924158  
Patricio Estrella                         0.891502   0.952583         0.894790  
Calamardo Tentáculos                      0.967265   0.968111         0.851644  
Arenita Mejillas                          0.962015   0.972120         0.900641  
Eugenio H. Cangrejo                       0.796723   0.896080         0.920664  
Sheldon J. Plankton                       0.922577   0.986733         0.969239  
Karen Plankton                            0.966088   0.984919         0.965252  
Perlita Cangrejo                          0.000000   0.873832         0.969504  
Sra. Puff                                 0.873832   0.0

In [142]:
dist_cos.describe()

Bob Esponja Pantalones Cuadrados  Patricio Estrella  \
count                         10.000000          10.000000   
mean                           0.792995           0.792471   
std                            0.281058           0.284439   
min                            0.000000           0.000000   
25%                            0.843257           0.827448   
50%                            0.870140           0.866312   
75%                            0.910436           0.921232   
max                            0.938609           0.966449   

       Calamardo Tentáculos  Arenita Mejillas  Eugenio H. Cangrejo  \
count             10.000000         10.000000            10.000000   
mean               0.811491          0.825837             0.763213   
std                0.291243          0.294148             0.276835   
min                0.000000          0.000000             0.000000   
25%                0.836419          0.856002             0.805378   
50%                0.918125          0.914029             0.849919   
75%                0.942399          0.958592             0.895034   
max                0.968111          0.972120             0.920664   

       Sheldon J. Plankton  Karen Plankton  Perlita Cangrejo  Sra. Puff  \
count            10.000000       10.000000         10.000000  10.000000   
mean              0.808482        0.837003          0.823532   0.845663   
std               0.299148        0.298973          0.294624   0.300812   
min               0.000000        0.000000          0.000000   0.000000   
25%               0.814255        0.901434          0.876827   0.879394   
50%               0.920092        0.943466          0.907039   0.958380   
75%               0.967443        0.961417          0.965070   0.971118   
max               0.986733        0.984919          0.969504   0.986733   

       Gary el Caracol  
count        10.000000  
mean          0.836007  
std           0.296353  
min           0.000000  
25%           0.896253  
50%           0.922411  
75%           0.964983  
max           0.969504

**¿Cuáles son los personajes más parecidos?**

Sheldon J. Plankton y Eugenio H. Cangrejo son los personajes más parecidos, ya que tienen una distancia de **0.687380**

**¿Cuáles son los personajes más diferentes?**

Sheldon J. Plankton y Sra. Puff son los personajes más diferentes, ya que tienen una distancia de **0.986733**